crates/ironrdp-async/src/connector.rs

```diff
@@ -177,3 +177,67 @@ where
 
     Ok(())
 }
+
+pub async fn single_connect_step<S>(
+    framed: &mut Framed<S>,
+    connector: &mut ClientConnector,
+    buf: &mut WriteBuf,
+) -> ConnectorResult<()>
+where
+    S: FramedWrite + FramedRead,
+{
+    buf.clear();
+    let written = single_connect_step_read(framed, connector, buf).await?;
+    single_connect_step_write(framed, buf, written).await
+}
+
+pub async fn single_connect_step_read<S>(
+    framed: &mut Framed<S>,
+    connector: &mut dyn Sequence,
+    buf: &mut WriteBuf,
+) -> ConnectorResult<Written>
+where
+    S: FramedRead,
+{
+    buf.clear();
+
+    if let Some(next_pdu_hint) = connector.next_pdu_hint() {
+        debug!(
+            connector.state = connector.state().name(),
+            hint = ?next_pdu_hint,
+            "Wait for PDU"
+        );
+
+        let pdu = framed
+            .read_by_hint(next_pdu_hint)
+            .await
+            .map_err(|e| ironrdp_connector::custom_err!("read frame by hint", e))?;
+
+        trace!(length = pdu.len(), "PDU received");
+
+        connector.step(&pdu, buf)
+    } else {
+        connector.step_no_input(buf)
+    }
+}
+
+async fn single_connect_step_write<S>(
+    framed: &mut Framed<S>,
+    buf: &mut WriteBuf,
+    written: Written,
+) -> ConnectorResult<()>
+where
+    S: FramedWrite,
+{
+    if let Some(response_len) = written.size() {
+        debug_assert_eq!(buf.filled_len(), response_len);
+        let response = buf.filled();
+        trace!(response_len, "Send response");
+        framed
+            .write_all(response)
+            .await
+            .map_err(|e| ironrdp_connector::custom_err!("write all", e))?;
+    }
+
+    Ok(())
+}
```

Comment:
question: Is it the place we agreed to put these functions? I kind of remember having commented about this in a previous PR.

Reply:
issue: I double checked the `framed` module. The function appears to be duplicated. Can you double check the diff again for that? Thank you!